<a href="https://colab.research.google.com/github/abhinavverma0907/SOC-2026-Encryption-and-Decryption/blob/main/Week_1/Assignment_2/Q2_Caesar_Decryption.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## SET-UP

In [19]:
alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
            'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
            'u', 'v', 'w', 'x', 'y', 'z', ' ', '.', ',', ';',
            ':', '/', '@']

N = len(alphabet)
char_to_idx = {ch: i for i, ch in enumerate(alphabet)}

def decrypt(text, shift):
    result = ""
    for ch in text:
        i = char_to_idx[ch]
        result += alphabet[(i - shift) % N]
    return result

print("Setup complete. Alphabet size:", N)

Setup complete. Alphabet size: 33


In [20]:
def frequency_analysis(ciphertext):
    freq = {}
    for ch in ciphertext:
        freq[ch] = freq.get(ch, 0) + 1

    sorted_freq = sorted(freq.items(), key=lambda x: x[1], reverse=True)

    print("Top 5 most frequent characters:")
    for ch, count in sorted_freq[:5]:
        print(f"  '{ch}' → {count} times")

    most_common = sorted_freq[0][0]
    print(f"\nMost common cipher char: '{most_common}'")

    shift = (char_to_idx[most_common] - char_to_idx[' ']) % N
    print(f"Predicted shift: {shift}")

    return shift

## DECRYPTING FIRST CIPHER TEXT - MARY'S LETTER

In [21]:
ciphertext1 = "dyvkrwwrzbcksvz:xkdyeck@bv@rbvukr:ukw/btvckz:kbvruz:vccks/dykgzdy/edkr:ukgzdyz:kdyvkbvr,;mkdyv:kcyr,,kzdksvkdz;vkd/kcvdkdyvkczhkxv:d,v;v:kd/kg/b.nkdr.z:xk/buvbmke@/:kdyvkrtt/;@,zcyz:xk/wkdyvzbkuvczx:mkzk;riksvkceuuv:,ikdbr:c@/bdvuk/edk/wkdyzck@,rtvmkr:ukdyrdkr,,ki/ebkw/btvckz:kdyvkcr;vkdz;vksvk/:kdyvkwzv,ukd/k;vvdk;vl"

print("=== MARY'S LETTER ===")
shift1 = frequency_analysis(ciphertext1)
print(f"\nDecrypted:\n{decrypt(ciphertext1, shift1)}")

=== MARY'S LETTER ===
Top 5 most frequent characters:
  'k' → 59 times
  'v' → 35 times
  'd' → 29 times
  'z' → 20 times
  ':' → 18 times

Most common cipher char: 'k'
Predicted shift: 17

Decrypted:
the affairs being thus prepared and forces in readiness both without and within the realm, then shall it be time to set the six gentlemen to work; taking order, upon the accomplishing of their design, i may be suddenly transported out of this place, and that all your forces in the same time be on the field to meet me.


## DECRYPTING SECOND CYPHER TEXT - BABINGTON POSTSCRIPT

In [22]:
ciphertext2 = "zkg/e,uksvkx,rukd/k.:/gkdyvk:r;vckr:ukaer,zdzvck/wkdyvkczhkxv:d,v;v:kgyztykrbvkd/krtt/;@,zcykdyvkuvczx:;v:dnkw/bkdyrdkzdk;riksvkzkcyr,,ksvkrs,vmke@/:k.:/g,vuxvk/wkdyvk@rbdzvcmkd/kxzfvki/ekc/;vkwebdyvbkrufztvk:vtvccrbikd/ksvkw/,,/gvukdyvbvz:"

print("=== BABINGTON POSTSCRIPT ===")
shift2 = frequency_analysis(ciphertext2)
print(f"\nDecrypted:\n{decrypt(ciphertext2, shift2)}")

=== BABINGTON POSTSCRIPT ===
Top 5 most frequent characters:
  'k' → 44 times
  'v' → 29 times
  'd' → 17 times
  '/' → 16 times
  'z' → 13 times

Most common cipher char: 'k'
Predicted shift: 17

Decrypted:
i would be glad to know the names and qualities of the six gentlemen which are to accomplish the designment; for that it may be i shall be able, upon knowledge of the parties, to give you some further advice necessary to be followed therein


## Q2 — Caesar Cipher Decryption

### Method: Frequency Analysis
In English text, **space** is the most common character (~13% of any normal text).
In a Caesar cipher every character shifts by the same fixed amount.
So the most frequent character in the ciphertext must be the encrypted version of space.

Steps:
1. Count frequency of every character in the ciphertext
2. Identify the most common one
3. Compute shift = (index of most common char) - (index of space) mod 33
4. Decrypt using that shift — no brute force needed

### Results

**Mary's Letter (shift = 17):**
the affairs being thus prepared and forces in readiness both without
and within the realm, then shall it be time to set the six gentlemen
to work...

**Babington Postscript (shift = 17):**
i would be glad to know the names and qualities of the six gentlemen
which are to accomplish the designment...

### Historical Note
Thomas Phelippes used the same shift (17) to forge the postscript —
sealing Mary Queen of Scots' fate and leading to her execution in 1587.